### ISSUE TITLE
Cannot fill in fields with pipes (|) in name-attribute

### ISSUE THREAD

**Issue by muhammedalialat** [minkphp/Mink#702]
I want to fill in field with pipes in the name a value, like

<code>And I fill in "a|b|c" with "1"</code>

The HTML form-element looks like:
```html
<input type="text" name="a|b|c" value="" />
```

The [Manipulator] splits the name-attribute.  May be the [NamedSelector] should escape the pipe-signs.

[Manipulator]: https://github.com/minkphp/Mink/blob/master/src/Selector/Xpath/Manipulator.php#L24
[NamedSelector]: https://github.com/minkphp/Mink/blob/master/src/Selector/NamedSelector.php#L244-L262

**Comment by aik099** [comment]
Agreed.

The `|` support wasn't initially there, when escaper code was created. I guess we forgot to update escaper code to support `|` when `Manipulator` class was added.

Actually escaping is done here: https://github.com/minkphp/Mink/blob/master/src/Selector/Xpath/Escaper.php#L27-L51

**Comment by aik099** [comment]
Not sure if `Manipulator` class will understand, that `\|` shouldn't be used for splitting.

**Comment by stof** [comment]
This indeed looks like a bug in the Manipulator when splitting the string. We will need a more complex implementation of the split.

**Comment by aik099** [comment]
Either we don't escape `|` (in escaper) or we don't honor `\|` (in manipulator) or both.

**Comment by aik099** [comment]
@muhammedalialat , you can submit PR with a fix & tests.

**Comment by stof** [comment]
@aik099 we don't need to escape the ``|`` as it is inside a XPath string. Escaping it would actually make the XPath invalid (it would look for a field named ``a\|b\|c``). The issue is the manipulator being too dumb

**Comment by aik099** [comment]
OK.

**Comment by muhammedalialat** [comment]
What about adding following line to https://github.com/minkphp/Mink/blob/master/src/Selector/PartialNamedSelector.php:

```php
$this->registerReplacement('%idOrNameMatch%', 'contains(./@id, %locator%) or contains(./@name, %locator%)');
```

Fields I'm dealing with looks more like
```html
<input type="text" name="wordOne|wordTwo|wordThree" value="" />
```

So I can use <code>And I fill in "wordOne" with "1"</code> to get the field filled. Or is there any wildcard sign to do something like <code>And I fill in "wordOne%wordTwo%wordThree" with "1"</code> with <code><strong>%</strong></code> as wildcard sign.
Or which way is easier to implement?

**Comment by stof** [comment]
@muhammedalialat adding a new feature (which will then have to be maintained forever) just to workaround a bug is a no-go. The bug should be fixed instead.

**Comment by gabel** [comment]
How should a solution work, if escapin `\|` is not the solution nor another feature?

The problem is that the pipe can be used in two context.

```gherkin
And I fill in "a|b|c" with "1"
```

Context 1
```html
<input type="text" name="a">
```

Context 2
````html
<input type="text" name="a|b|c">
```

Both contexts should be filled with the sentence...

**Comment by aik099** [comment]
You're not matching (what `Context 1` implies), but filling (what `Context 2` implies).

**Comment by gabel** [comment]
I've added a proposal for a fix that fulfills the tests. If the fix is OK, I'll add tests.

**Comment by claudiu-cristea** [comment]
I'm getting the same when searching for links that have pipe (`|`) in text. The problem is that `|` is used as operator in XPath expressions, so we need to escape `|` that is a locator. But how to escape it and where? I tried with `\|` but doesn't work.

**Comment by stof** [comment]
the selector already takes care of the escaping. The issue is that our XPath Manipulator used to prepend the element xpath is not smart enough to handle XPath expression where ``|`` appears inside strings (it explodes the strings blindly instead of being aware of XPath strings)

**Comment by claudiu-cristea** [comment]
@stof, Just to test if this is the cause, in Manipulator I tried to replace `\|` with an arbitrary placeholder and then replaced back, later, after imploding to avoid exploding on that pipe. First, the pipe is not escaped in that point it's just `|`, not `|`. Then I did the same with `|` but no luck.

**Comment by claudiu-cristea** [comment]
PR #718 contains a unit test that reveals the bug.

**Comment by stof** [comment]
Well, the point is there is **no** reason to escape the ``|`` in XPath when being inside a string. So this is why your "hack" does not work. and if you replace all ``|``, you break the whole feature

**Comment by claudiu-cristea** [comment]
@stof, yes, I agree. The phpunit from #718 show it clearly.

**Comment by stof** [comment]
Fixed in #720
